apps/api-user/src/modules/user/user.service.ts

```diff
@@ -0,0 +1,395 @@
+  ) { }
+
+  // DONE
+  async isUserBanned(userId: string): Promise<boolean> {
+    const user = await this.prisma.user.findUnique({
+      select: {
+        bannedUntil: true,
+        isBanned: true,
+      },
+      where: { id: userId },
+    });
+
+    if (!user) {
+      throw new Error("User not found");
+    }
+
+    if (!user.isBanned) {
+      return false;
+    }
+
+    if (!user.bannedUntil) {
+      return true;
+    }
+
+    return user.bannedUntil > new Date();
+  }
+
+  // DONE
+  async banUser(id: string, dto: BanUserDto, adminId: string): Promise<void> {
+    const user = await this.prisma.user.findUnique({ where: { id } });
+
+    if (!user) throw new NotFoundException('User not found');
+
+    await this.prisma.user.update({
+      where: { id },
+      data: {
+        isBanned: true,
+        bannedOn: new Date(),
+        bannedUntil: dto.bannedUntil ? new Date(dto.bannedUntil) : null,
+        banReason: dto.banReason ?? null,
+        bannedById: adminId,
+      },
+    });
+  }
+
+  // DONE
+  async unbanUser(id: string): Promise<void> {
+    const user = await this.prisma.user.findUnique({ where: { id } });
+
+    if (!user) throw new NotFoundException('User not found');
+
+    if (!user.isBanned && !user.bannedUntil && !user.bannedOn && !user.banReason && !user.bannedById) {
+      return;
+    }
+
+    await this.prisma.user.update({
+      where: { id },
+      data: {
+        isBanned: false,
+        bannedOn: null,
+        bannedUntil: null,
+        banReason: null,
+        bannedById: null,
+      },
+    });
+  }
+
+  // pay attention to related records while deleting, same with soft delete
+  async hardDeleteUser(userId: string): Promise<void> {
+    const exists = await this.prisma.user.findUnique({
+      select: { id: true },
+      where: { id: userId },
+    });
+
+    if (!exists) {
+      throw new NotFoundException('User not found');
+    }
+  }
+
+  // this is complicated, leave for now
+  async softDeleteUser(): Promise<PublicUserDto> {
+    //dla usera i admina, jak snapchat
+    return;
+  }
+
+  // DONE 
+  async getUserDetails(userId: string): Promise<PublicUserDto> {
+    const user = await this.prisma.user.findUnique({
+      select: {
+        _count: {
+          select: {
+            followers: true,
+            following: true,
+          },
+        },
+        createdAt: true,
+        email: true,      // keep only if public
+        id: true,
+        username: true,
+        bio: true,        // add to match PublicUserDto
+      },
+      where: { id: userId },
+    });
+
+    if (!user) {
+      throw new NotFoundException('User not found');
+    }
+
+    return {
+      createdAt: user.createdAt,
+      email: user.email,
+      followersCount: user._count.followers,
+      followingCount: user._count.following,
+      id: user.id,
+      username: user.username,
+      bio: user.bio ?? undefined,
+    };
+  }
+
+  // pagination
+  async findAll(): Promise<PublicUserDto[]> {
```

Comment:
You will need to create pagination in this way (it's not how it should look, but example)

```suggestion
async findAll(page = 1, limit = 20): Promise<{ users: PublicUserDto[], total: number }> {
  const [users, total] = await Promise.all([
    this.prisma.user.findMany({
      skip: (page - 1) * limit,
      take: limit,
      // other parameters
    }),
    this.prisma.user.count()
  ]);

  return { users: mappedUsers, total };
}
```

Also, you can create util/common class/method to comfortably create pagination, like extract from query data (page/limit), like build response in pagination style (with meta part (total, currentPage, nextPage etc))

Reply:
done